### PR TITLE
[Feature Anywhere] Add view events flyout

### LINKED ISSUE #3155: [Feature Anywhere] Support a 'View Events' flyout for event details
## Overview

Coming from #2880 

The project involves supporting a new flyout with a holistic view on all event annotations partitioned by plugin, and then by plugin resource, for a particular visualization. To achieve this, we need to create a separate annotation line chart for each unique `pluginResourceId` value found in the set of events, and map them to an individual annotation line chart that will all need to be stacked on a consistent timeline.

This flyout will need to live in core Dashboards, and support logic such that any associated plugin resources that are associated with a vis can be cleanly organized and viewable in the flyout.

We will support a soft limit of 10 plugin resources per visualization. If the user changes that, scrolling may be shown in the flyout body. We will not support pagination currently.

### Opening the flyout
There are 2 ways to open the flyout - via a 'View events' button in the vis options panel, or clicking on an event point in the chart itself. 
To open via 'View events' button, we can create a trigger and corresponding action using the embeddable as context, similar to the existing way actions are triggered from different menu actions.
To open by clicking on the chart, we can use the saved object ID that is passed & persisted in the chart as context.

### Handling context
When the flyout opens, we want to use the same context given to the source vis embeddable (query / filters / time range), in order to render the new vis embeddables and show the same data. We can use the data plugin's `queryService` to do this, which fetches the global context.

### Handling refresh
By default, we don't want to have the charts auto-refresh in the flyout (if enabled at the dashboard level). We can manually turn this off by calling `updateInput()` on the embeddable with an updated `refreshConfig`. If the user wants to do a manual refresh, we have a button at the top of the flyout where, when clicked, calls `reload()` on each visible embeddable.

### Fetching `VisLayer`s
Given some embeddable, we need to be able to access its latest set of `VisLayer`s so the flyout can process them accordingly. To do this, we can add a `visLayers` field in the visualize embeddable. It can be set in `updateHandler()`, which is triggered upon any input changes to the embeddable where they are initially fetched.

### Generating `VisLayer` embeddable charts
We can easily generate the source vis embeddable by creating a new one from its saved object ID and passing in the identical global context, as mentioned above. To generate the individual charts below (one per `VisLayer`), we need to pull out individual time series data and display them in separate charts.

Because there is no easy way to extract the source datatable from the source vis embeddable to filter out data and generate a new chart, we need to find a different way. If we take on the perspective that each `VisLayer` chart is the same as the source chart but just with filtered data, this makes things simpler. We can introduce a new optional `visLayerResourceIds` field in the visualize embeddable, such that we will only show events / layers that are coming from the set of `visLayerResourceIds`. In the context of this flyout, we can just re-generate an embeddable for each `VisLayer`, passing in a single `visLayerResourceId`.

This gives us a very flexible advantage and opens the door for many interesting ways we may want to overlay data on the charts in the future, such as being able to toggle on/off all events coming from one particular plugin, or a subset of plugins, or a subset within a plugin, etc.

### Adding fields to the `VisLayer`
There is some additional fields we need to persist within a `VisLayer` to properly render the flyout. This includes the resource name, url, and resource type. Those are added separately as part of #3374.

Example mockup (not finalized):
Screen Shot 2022-12-29 at 9 49 12 AM

~~Open questions:~~
~~1. Should there be a default limit on the number of resources viewable?~~
~~2. If no limit, how should scrolling and/or searching be handled to access all potential plugin resource line charts?~~

### LINKED ISSUE #3485: [BUG] Vislib to Vega-lite migrator, fix chart overflow issue
**Describe the bug**

For charts that are being rendered with vega-lite instead of vislib, the chart has an overflow error where you need to scroll in the chart to view the entire chart as there is some css styling problem that doesnt size the chart correctly when displaying on a visualize embeddable. This is related to the #2880 feature and came from the #3106 PR.

**To Reproduce**
Create a line chart that is rendered by vega-lite

**Expected behavior**
Have the chart fix in the embeddable object without the need to scroll.

### LINKED ISSUE #3580: [Feature Anywhere] General error handling of `VisLayer` errors
The `VisLayer` error-related interfaces will be defined in #3267. When that is merged, this can be worked on.

Some notes on how we want functionality:
- for any errors (security or general), show a general toast for each plugin. users can click a button to open up a modal to view more details on the error(s) (error type & error message) for the affected plugin resource(s)

### PR DESCRIPTION
**Pull request by ohltyler**
Signed-off-by: Tyler Ohlsen <ohltyler@amazon.com>

Currently this PR has many dependent ones that will need to be merged first. This includes:

- [x] #3106 
- [x] #3109 
- [x] #3131 
- [x] #3374
- [x] #3145 
- [x] #3267 
- [x] #3317 

These all involve finalizing the rest of the chart render workflow, and the `VisLayer` design.

When these have been merged, we can finish the following items for this flyout: 
- [x] final parsing of the `VisLayer`s to show the partitioned data correctly
- [x] custom chart functionality, including being able to (1) not show axes & labels, (2) not show any data, and (3) removing click functionality. (1) and (2) will mostly depend on #3106 & #3145, and (3) will mostly depend on #3317 
- [x] error handling of failed `VisLayer`s which will be dependent on #3267

---

## What can be reviewed now
- all files under `view_events_flyout/components/`
- all files under `vis_type_vega` plugin
- all files under `vis_type_vislib` plugin

---

## Description
This PR adds the 'View events' flyout which will open via an event click on a chart with `VisLayer`s enabled, or in the vis options menu via `View events` button. It shows the base vis with the event annotations, and below it a breakdown of each set of events produced by each plugin resource. It is categorized by plugin resource type as well.

**Technical details, more information, and example mocks can be found in the related issue #3155** 

## Validation
To test the feature:
- clone locally (no extra plugins needed)
- create an eligible line visualization (for eligibility see #3687 until docs are finished)
- embed the visualization on some dashboard
- update the timestamps in the [list of test VisLayers](https://github.com/ohltyler/OpenSearch-Dashboards/blob/view-events-2/src/plugins/visualizations/public/embeddable/visualize_embeddable.ts#L572) to be within the time range of your visualization
- events should appear, and can be clicked on the chart directly or via "view events" button in the options dropdown for the embeddable panel, as seen in sample video below

### Components
`ViewEventsFlyout` - the base component where all embeddables are created from the source vis embeddable. Persists the event visualizations by type via a map mapping resource type -> list of embeddables

`DateRangeItem` - contains the selected date range and refresh button. Hooks into the `reload()` function defined in `ViewEventsFlyout` to reload all of the embeddables when clicked.

`BaseVisItem` - the re-created base vis embeddable containing all of the events

`EventsPanel` - the base component to show all of the event annotation charts. Contains a set of `PluginEventsPanel`s 

`PluginEventsPanel` - contains a set of `EventVisItem`s for a particular plugin resource type

`EventVisItem` - contains an event vis embeddable, including the resource name, url, and event count

`EventVisItemIcon` - contains the icon with the event count or the error icon with error message details, if applicable.

`TimelinePanel` - contains the empty embeddable. Used to re-create the timeline and is a static component on the bottom of the flyout

All of the embeddables are using the `EmbeddablePanel` component, which takes care of individual component lifecycles for us, including the rendering, refreshing, and destroying. This is fetched from the embeddable service function's `getEmbeddablePanel()`.

### The `VisAugmenterEmbeddableConfig`
The flyout requires rendering essentially the same visualization, but with various visual changes and certain data filtered out; for example, showing only data from one `VisLayer`, showing just the x-axis, hiding the y axes, adding horizontal padding, etc. To account for these changes, we need to configure this at the top-level such that it is taken into account when generating the high-level vega-lite spec at the visualize_embeddable level, all the way down to the compiled vega spec and instantiated vega view. To do this, we add the `VisAugmenterEmbeddableConfig`. This is a high-level interface (independent of charting libraries/details) that is used to specify these necessary configurations:
- `visLayerResourceIds` - the list of `VisLayers` with particular plugin resource IDs we want to show on the chart
- `inFlyout` - whether we are rendering this embeddable within the view events flyout or not
- `flyoutContext` - whether we are rendering a base vis (the original vis), event vis (the individual VisLayer charts), or timeline vis (the bottom vis only showing the x-axis / timeline)
- `leftValueAxisPadding` - whether we need to add padding to the left of the value axis for alignment
- `rightValueAxisPadding` - whether we need to add padding to the right of the value axis for alignment

Demo video:
[demo-new.webm](https://user-images.githubusercontent.com/62119629/234130551-714223d9-440e-4cb0-9dc5-a68ddfbae2b3.webm)

 
### Issues Resolved
Closes #3155 
Closes #3580 
Closes #3485 
 
### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by ohltyler** [comment]
Some more details on how we will handle error handling (dependent on #3267).
- if there is security read permissions (either fetching resource or fetching data) we return a security error. this resource (and likely the plugin itself) will not be visible anywhere in the feature (the resource lists, view events page, etc.).
- if there is general failure to fetch a resource, show one toast per plugin, with a button to open a modal and view the errors at per-resource granularity for more information
- if there is failure to fetch all resources for a particular plugin (empty plugin list), we should hide the title altogether in view events page

**Comment by ohltyler** [comment]
While the final items (custom chart functionality and error handling) are currently blocked, the bulk of this PR is ready and able to be reviewed. Please read PR description and corresponding for some more context, before starting review.

**Comment by joshuarrrr** [comment]
@ohltyler Set this to draft until it's ready for review by a feature anywhere dev.

**Comment by ohltyler** [comment]
> @ohltyler Set this to draft until it's ready for review by a feature anywhere dev.

Given the size of this PR, it can definitely speed things up if there is folks from OSD to start reviewing concurrently, and earlier. The major bulk of all changes are fully ready for review. I would call out any of the upcoming changes.

File: src/plugins/vis_augmenter/public/vega/helpers.ts
Comment (lezzago) [inline src/plugins/vis_augmenter/public/vega/helpers.ts:418]: Can we add a comment that we choose these specific static values to make the alignment right?

**Comment by ohltyler** [comment]
I've rebased with the approximate changes of the click handling PR. I've got it integrated and working as expected.

Minor changes remaining:
- [ ] disabling clicking when the flyout is open (need to persist global flyout state)
- [ ] contents of custom tooltips - may be entirely covered via #3317 
- [ ] resolve options menu bug so the option is visible again (seems to be after rebasing with latest 3.0 changes coming from OSD core?)

File: src/plugins/vis_augmenter/public/vega/helpers.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/vega/helpers.ts]: done

**Comment by joshuarrrr** [comment]
@ohltyler I'll go ahead and review this, but looks like there are a few conflicts that will need to be resolved.

**Comment by ohltyler** [comment]
> @ohltyler I'll go ahead and review this, but looks like there are a few conflicts that will need to be resolved.

Yes, I've had to prematurely pull some uncommitted changes from other PRs to ensure integration was working as expected, hence the weird conflicts. I will clean these up once the final dependent PR #4120 is merged.

File: src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx
Comment (joshuarrrr) [inline src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx:240]: Why did we need to add border and shadow props to the embeddable panel for the anywhere feature?

File: src/plugins/ui_actions/public/triggers/external_action_trigger.ts
Comment (joshuarrrr) [inline src/plugins/ui_actions/public/triggers/external_action_trigger.ts:35]: nit - no need to change, but I think the naming here is a little overly generic compared to the description. It's also a little weird to me that the title hard codes the interaction - I don't see why an external action (such as opening a flyout requires a single click, as opposed to a double click or some hotkey combo.

File: src/plugins/vis_type_vega/public/vega_view/vega_base_view.js
Comment (joshuarrrr) [inline src/plugins/vis_type_vega/public/vega_view/vega_base_view.js:274]: I see the TODO is removed, but this now seems like a magic number - it would be hard to know in the future whether it can be safely changed and how it was determined.

File: src/plugins/vis_type_vega/public/expressions/helpers.ts
Comment (joshuarrrr) [inline src/plugins/vis_type_vega/public/expressions/helpers.ts:61]: Why do we need to explicitly calculate this offset based on legend position?

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (joshuarrrr) [inline src/plugins/visualizations/public/embeddable/visualize_embeddable.ts:529]: @ohltyler Was your plan to replace the hardcoded values in a subsequent commit to this PR, or in a separate follow-up PR?

**Review by joshuarrrr — COMMENTED** [review]
A couple of minor questions to assist my understanding. I've gotten through changes to existing plugins - there's a lot of changes to vis_augmenter I haven't reviewed yet, but those are less critical because it's a brand new component that's much less likely to introduce regressions. I'll finish a quick review of those files in the morning.

File: src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx
Comment (ohltyler) [reply src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx:240]: It was needed so we can have them removed/invisible in the view events flyout embeddables. Otherwise there was unnecessary lines and borders in the flyout that we didn't want and weren't part of the UX mocks.

File: src/plugins/ui_actions/public/triggers/external_action_trigger.ts
Comment (ohltyler) [reply src/plugins/ui_actions/public/triggers/external_action_trigger.ts:35]: Apologies, this action was actually added in #3777 , they show as changes here since I hadn't rebased. You may ignore that part, but if you have more concerns for that, I can try to address that after rebasing.

File: src/plugins/vis_type_vega/public/vega_view/vega_base_view.js
Comment (ohltyler) [reply src/plugins/vis_type_vega/public/vega_view/vega_base_view.js:274]: I'll add a comment on the reasoning here. This is the pixel height of the event canvas such that there is sufficient but minimal space to render the event marks. It is correlated to the mark size.

File: src/plugins/vis_type_vega/public/expressions/helpers.ts
Comment (ohltyler) [reply src/plugins/vis_type_vega/public/expressions/helpers.ts:61]: When the legend is at a vertical position (top or bottom), it shifts the chart canvas up or down to compensate, and can cause it to get cut off in the panel. I believe the root cause is that the vega-lite autosize has some issue of not compensating for this within the panel. So to fix this, i set the offset to 0 so the canvas doesn't move to guarantee it does not get cut off.

Comment (ohltyler) [reply src/plugins/vis_type_vega/public/expressions/helpers.ts:61]: The issue does not arise when the legends are horizontal (left/right), so I believe it is a vega-lite bug.

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (ohltyler) [reply src/plugins/visualizations/public/embeddable/visualize_embeddable.ts:529]: I will remove as a final commit to this PR after a bit more testing. Let me rebase and check functionality, then I'll remove.

File: src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx
Comment (joshuarrrr) [reply src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx:240]: got it - do we actually need to use panels within the flyout? I seem to remember that there are other options for embeddables without using the panel wrapper, which sounds closer to what we want.

Comment (ohltyler) [reply src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx:240]: Ah right - I learned about this yesterday actually, let me look into it: https://github.com/opensearch-project/alerting-dashboards-plugin/pull/553

**Comment by ohltyler** [comment]
@joshuarrrr rebased without making any changes. Follow-up commits will be to address comments.

File: src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx
Comment (ohltyler) [reply src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx:240]: Using the renderer seems to break the charts from being visible in the flyout. My assumption is the way that we have the panels wrapped within flex items doesn't work the same for the renderer and causes sizing or positioning issues.

For now, prefer to leave as-is with the panel wrapper. It ensures consistent look/feel to the vis panels as seen in the dashboard, and the modification to use the renderer may require refactoring and effort to get the sizing & positioning to work as expected.

File: src/plugins/vis_type_vega/public/vega_view/vega_base_view.js
Comment (ohltyler) [reply src/plugins/vis_type_vega/public/vega_view/vega_base_view.js]: added

File: src/plugins/vis_type_vega/public/expressions/helpers.ts
Comment (ohltyler) [reply src/plugins/vis_type_vega/public/expressions/helpers.ts:66]: added a comment about this

File: src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx
Comment (joshuarrrr) [reply src/plugins/embeddable/public/lib/panel/embeddable_panel.tsx:240]: That seems fine. It's an aspect of embeddables I'd love for us to get right so it's easy for plugins to embed in either an empaneled or clean state. I'll open a tracking issue.

File: src/plugins/vis_type_vega/public/expressions/helpers.ts
Comment (joshuarrrr) [reply src/plugins/vis_type_vega/public/expressions/helpers.ts:66]: Thanks - this is good for now, although we'll probably eventually want a more robust solution to vega responsive sizing in general. I appreciate all the research and improvements you've made to help us get this far.

Comment (ohltyler) [reply src/plugins/vis_type_vega/public/expressions/helpers.ts:66]: Definitely definitely agreed. Overall I'd say the existing integration of vega lite autosizing within the embeddable panels is a big mess. There's too many places that needs to be touched to get it to work properly - vega-lite spec, modifying compiled vega spec, modifying the vega view

**Review by joshuarrrr — COMMENTED** [review]
Overall this looks great and was clearly a ton of work. I mostly have some minor questions and comments, but no blockers.

File: src/plugins/vis_augmenter/public/constants.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/constants.ts:7]: I believe so, I had removed it since with all of the concurrent OUI going on it's not needed for now. If it's decided later on we want more consistency (including the colors used in the regular chart data), we can modify.

File: src/plugins/vis_augmenter/public/plugin.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/plugin.ts:61]: Let me get back to you on that. This was from conflict resolutions too. We are finalizing where we need these settings, since there's some tricky dependency situations to handle when consumed by plugins, but using plugin functionality in other plugins (like Visualizations) that aren't dependent on the plugins.

File: src/plugins/vis_augmenter/public/types.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/types.ts:82]: Yeah, was trying to strike a right balance here, but given the uncertainty in what this may look like later on I leaned more on the specific side for now.

File: src/plugins/vis_augmenter/public/view_events_flyout/components/utils.tsx
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/view_events_flyout/components/utils.tsx:25]: Definitely agreed. I think this is a pretty hacky solution. I'm still working on testing out a bunch of different chart configs to see if I can break it.

I think this is a good argument towards the chart abstraction needs. Especially with vega where we have to touch it at so many different places - vega-lite spec, compiled vega spec, and the generated vega view.

File: src/plugins/vis_augmenter/public/view_events_flyout/components/view_events_flyout.tsx
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/view_events_flyout/components/view_events_flyout.tsx:87]: I haven't ran into that case yet - typically I've relied on the vis embeddables to handle those scenarios - for example, if there is some error within the embeddable it will just render an error embeddable with the error message. I've reproduced this a bunch of times. This is why i only do undefined checks here.

I'll update after more plugin integration testing and see if we can trigger an infinite loading case.

File: src/plugins/vis_augmenter/public/view_events_flyout/components/view_events_flyout.tsx
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/view_events_flyout/components/view_events_flyout.tsx:95]: I had done that to preserve the flyout header height to being static, so that it doesn't change sizes when transitioning from loading -> loaded state. I'll switch to the text loading component and "Error fetching events" options so it's more informative.

File: src/plugins/visualizations/public/embeddable/visualize_embeddable.ts
Comment (ohltyler) [reply src/plugins/visualizations/public/embeddable/visualize_embeddable.ts]: I've removed this

File: src/plugins/vis_type_vislib/public/line_to_expression.ts
Comment (ohltyler) [reply src/plugins/vis_type_vislib/public/line_to_expression.ts:46]: Thanks - fixed.

File: src/plugins/vis_augmenter/public/view_events_flyout/components/styles.scss
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/view_events_flyout/components/styles.scss:25]: Will add.

File: src/plugins/vis_augmenter/public/plugin.ts
Comment (ohltyler) [reply src/plugins/vis_augmenter/public/plugin.ts:61]: We can remove from setup(). The dependency issue is fixed for now in #4139

Comment (ohltyler) [reply src/plugins/vis_augmenter/public/plugin.ts:60]: removed

File: src/plugins/vis_augmenter/public/constants.ts
Comment (joshuarrrr) [reply src/plugins/vis_augmenter/public/constants.ts:7]: Opened https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4143

File: src/plugins/vis_augmenter/public/view_events_flyout/components/utils.tsx
Comment (joshuarrrr) [reply src/plugins/vis_augmenter/public/view_events_flyout/components/utils.tsx:25]: Yeah - I know we did things a bit out of order for this feature, but I'd love your help to debrief and redesign some of this stuff once we ship.

File: src/plugins/vis_augmenter/public/view_events_flyout/components/styles.scss
Comment (joshuarrrr) [inline src/plugins/vis_augmenter/public/view_events_flyout/components/styles.scss:7]: thanks, this will be very useful when we refactor some of this styling!

**Review by joshuarrrr — APPROVED** [review]
Thanks @ohltyler for all the fixes. I'm going to mark as approved, but I'll need to reapprove after we update the snapshots.

**Comment by joshuarrrr** [comment]
@ohltyler There are still some flyout unit test failures - maybe something got messed up in the revision commits: https://github.com/opensearch-project/OpenSearch-Dashboards/actions/runs/5085933677/jobs/9139910721?pr=3415

**Review by lezzago — APPROVED** [review]

**Review by manasvinibs — APPROVED** [review]

**Comment by ohltyler** [comment]
@manasvinibs We are still looking to get a review from @ashwin-pc who has more context on the PR - let's hold off on merging until that is done

**Comment by ohltyler** [comment]
Discussed with @ashwin-pc - will look into the possibility of removing the dependency on the embeddables plugin as well as the added config to `VisualizeEmbeddable`, and instead render the charts using `ReactExpressionRenderer`, similar to what `VisBuilder` plugin does. This can simplify the logic needed to produce such charts, and allow us to control the input config that we've needed to add to `VisualizeEmbeddable` (the `VisAugmenterEmbeddableConfig`).

**Comment by ohltyler** [comment]
> 

Will follow up in a separate PR for this.

**Pull request merged**
